README.md
# react-multi-bar-slider
Slider component with multiple bars for React.

[Demo repo](https://github.com/rafaelklaessen/react-multi-bar-slider-demo)

![Multi slider screenshot](https://github.com/rafaelklaessen/react-multi-bar-slider/raw/master/screenshots/screenshot.png "Screenshot of the multi slider demo")

## Install
`yarn add react-multi-bar-slider` or `npm install --save react-multi-bar-slider`

## Usage
See the [demo repo](https://github.com/rafaelklaessen/react-multi-bar-slider-demo).

### Basic example
```javascript
import MultiSlider, { Progress } from 'react-multi-bar-slider';

export default class App extends Component {
  state = {
    progress: 10
  };

  handleSlide = newProgress => this.setState({ progress: newProgress });

  render() {
    return (
      <MultiSlider onSlide={this.handleSlide}>
        <Progress color="green" progress={this.state.progress} />
        <Progress color="purple" progress={45} />
      </MultiSlider>
    );
  }
}
```

### Advanced example
```javascript
import MultiSlider, { Progress, Dot } from 'react-multi-bar-slider';

export default class App extends Component {
  state = {
    progress: 10
  };

  handleSlide = newProgress => this.setState({ progress: newProgress });

  render() {
    return (
      <MultiSlider
        width={600}
        height={20}
        slidableZoneSize={40}
        backgroundColor="bisque"
        equalColor="blue"
        style={{ marginBottom: 40 }}
        onSlide={this.handleSlide}
        onDragStart={progress => console.log(`Started dragging: ${progress}%`)}
        onDragStop={progress => console.log(`Stopped dragging: ${progress}%`)}
        roundedCorners
        reversed
      />
        <Progress color="green" progress={this.state.progress}>
          <Dot color="grey" />
        </Progress>
        <Progress color="purple" progress={45}>
          <Dot color="grey" />
        </Progress>
      </MultiSlider>
    );
  }
}
```

## Props
`*` = Required

### `MultiSlider`
Prop | Description | Type | Default
---- | ----------- | ---- | -------
`width` | Width of the slider | `number` or `string` | `100%`
`height` | Height of the slider | `number` or `string` | `14px`
`slidableZoneSize` | Size of the zone in which sliders can be dragged | `number` or `string` | `7px`,
`backgroundColor` | Background color of the slider | `string` | `#EEEEEE`
`equalColor` | Color of all bars when their values are equal | `string` |
`style` | Custom style for the slider <br> *Signature:* <br> `function(props: object) => object` | `object` or `function` | `{}`
`onSlide` | Callback that is fired when the progress was set <br> *Signature:* <br> `function(progress: number) => void` | `function` |
`onDragStart` | Callback function that is fired when the slider has begun to move <br> *Signature:* <br> `function(progress: number) => void` | `function` |
`onDragStop`| Callback function that is fired when the slide has stopped moving <br> *Signature:* <br> `function(progress: number) => void` | `function` |
`roundedCorners` | When set to `true`, the slider has rounded corners | `bool` | `false`
`reversed` | When set to `true`, the slider is reversed | `bool` | `false`
`readOnly` | When set to `true`, the slider can't be updated | `bool` | `false`
`children*` |  The progress bars that are shown in the slider (or any other children) | `node` |

All other props (not documented here) will be passed on to the root element.

### `Progress`
Prop | Description | Type | Default
----- | ----------- | ---- | -------
`color*` | Color of the progress bar | `string` |
`progress*` | Progress of the progress bar | `number` |
`style` | Custom style for the progress bar <br> *Signature:* <br> `function(props: object) => object` | `object` or `function` |
`children` | The slider dot (or any other children) | `node` |

All other props (not documented here) will be passed on to the root element.

### `Dot`
Prop | Description | Type | Default
----- | ----------- | ---- | -------
`width` | Width of the dot | `number` or `string` | `50` when the dot has an icon, `28` if not
`height` | Height of the dot | `number` or `string` | `50` when the dot has an icon, `28` if not
`color` | Color of the dot | `string` |
`icon` | URL of the image to use as dot icon | `string` |
`style` | Custom style for the dot <br> *Signature:* <br> `function(props: object) => object` | `object` or `function` |
`iconStyle` | Custom style for the dot icon <br> *Signature:* <br> `function(props: object) => object` | `object` or `function` |
`children` | Children of the dot | `node` |

All other props (not documented here) will be passed on to the root element.

#### How custom styles work
When a function is passed to a `style` prop rather than an object, it is expected to return an object.
The style function will be called with all props that that component has (except for the `style` and `children` props and any internal callbacks). The return value of the function will be used as style (see the [demo repo](https://github.com/rafaelklaessen/react-multi-bar-slider-demo) for an example).

## Caveats
When the `MultiSlider`, `Progress` or `Dot` components are wrapped with glamorous (eg `glamorous(MultiSlider)`), glamorous breaks. I don't know how to fix this, so don't wrap them :)


CHANGELOG.md
# Version 1.3.1
Released 2019-05-18

- Fixed issues with touch events

# Version 1.3.0
Released 2019-05-14

- Added support for touch events

# Version 1.2.1
Released 2018-09-08

- Added `react-emotion` to `peerDependencies`

# Version 1.2.0
Released 2018-09-08

- Migrated from `glamorous` to `emotion`
- Dragging now only activates when the left mouse button is clicked

# Version 1.1.0
Released 2018-03-28

- Fixed dependency versions in `package.json`
- Moved from `PureComponent` to `Component`
- Added `onDragStart` and `onDragStop` callbacks
- Updated README

# Version 1.0.1
Released 2018-03-23

- Removed public folder

# Version 1.0.0
Released 2018-03-23

- Initial release


src/MultiSlider.js
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';

export const transition = 'all 450ms cubic-bezier(.23, 1, .32, 1) 0ms';

export const ProgressSlider = glamorous.div({
  height: 14,
  boxSizing: 'border-box',
  transition
}, ({ readOnly, width, height, backgroundColor }) => ({
  width,
  height,
  backgroundColor,
  cursor: readOnly ? 'auto' : 'pointer'
}));

const Dot = glamorous.span({
  position: 'absolute',
  display: 'block',
  zIndex: 5,
  borderRadius: '50%'
}, ({ right, hasIcon, width, height, color, noTransition }) => ({
  top: hasIcon ? 0 : '50%',
  right: `${right || 0}%`,
  transform: hasIcon ? 'translateX(-50%)' : 'translate(50%, -50%)',
  width: hasIcon ? 0 : width,
  height: hasIcon ? 0 : height,
  backgroundColor: hasIcon ? 'transparent' : color,
  transition: noTransition ? 'none' : transition
}));

const DotIcon = glamorous.img({
  position: 'absolute',
  transform: 'translateX(-50%)',
  userSelect: 'none'
}, ({ width, height }) => ({
  width,
  height
}));

export const Progress = glamorous.div({
  position: 'absolute',
  top: 0,
  right: 0
}, ({ color, progress, height, equal, equalColor, zIndex, noTransition }) => ({
  width: `${progress || 0}%`,
  height,
  backgroundColor: equal && equalColor ? equalColor : color,
  zIndex,
  transition: noTransition ? 'none' : transition
}));

export const SlidableZone = glamorous.div({
  position: 'absolute',
  width: '100%'
}, ({ size, sliderCount }) => ({
  top: -size,
  bottom: -size,
  zIndex: sliderCount
}));

const limitProgress = progress => Math.max(Math.min(progress, 100), 0);

export default class MultiSlider extends Component {
  static propTypes = {
    sliders: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string.isRequired,
        progress: PropTypes.number.isRequired,
        dot: PropTypes.oneOfType([
          PropTypes.shape({
            color: PropTypes.string,
            icon: PropTypes.string,
            style: PropTypes.object
          }),
          PropTypes.bool
        ])
      }).isRequired
    ).isRequired,
    activeSlider: PropTypes.number,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    slidableZoneSize: PropTypes.number,
    backgroundColor: PropTypes.string,
    equalColor: PropTypes.string,
    sliderStyle: PropTypes.object,
    onSlide: PropTypes.func.isRequired,
    readOnly: PropTypes.bool
  };

  static defaultProps = {
    activeSlider: 0,
    width: '100%',
    height: 14,
    slidableZoneSize: 7,
    backgroundColor: '#EEEEEE',
    sliderStyle: {},
    readOnly: false
  };

  state = { mouseDown: false };

  handleSliderClick = (e) => {
    const { sliders, activeSlider, onSlide, readOnly } = this.props;

    if (readOnly) return;

    const sliderEl = this.sliderElement(e);
    const newProgress = this.progressFromMousePos(e, sliderEl);

    const oldProgress = sliders[activeSlider].progress;
    if (newProgress === oldProgress) return;

    onSlide(newProgress);
  };

  handleSliderMouseDown = () => this.setState({ mouseDown: true });
  handleSliderMouseUp = () => this.setState({ mouseDown: false });

  handleSliderMouseMove = (e) => {
    if (!this.state.mouseDown) return;
    this.handleSliderClick(e);
  };

  sliderElement = (e) => {
    const sliderEl = e.target;

    // Make sure we select the slider element
    if (sliderEl.classList[1] === 'icon') {
      return sliderEl.parentNode.parentNode;
    }
    if (sliderEl.classList[1] === 'dot' || sliderEl.classList[1] === 'progress') {
      return sliderEl.parentNode;
    }
    return sliderEl;
  };

  progressFromMousePos = (e, sliderEl) => {
    const boundingRect = sliderEl.getBoundingClientRect();
    const mouseFromLeft = e.pageX - boundingRect.left;
    const progressFromLeft = mouseFromLeft / boundingRect.width;
    // Return progress in percents
    return limitProgress(Math.round((1 - progressFromLeft) * 100));
  };

  sortSliders = () => {
    const sliders = [...this.props.sliders];
    return sliders.sort((a, b) => b.progress - a.progress);
  };

  slidersEqual = () => {
    const sliders = this.props.sliders;
    return sliders.reduce((acc, slider, i) => {
      if (!acc) return acc;
      if (!sliders[i - 1]) return true;
      if (slider.progress === sliders[i - 1].progress) return true;
      return false;
    }, true);
  };

  render = () => {
    const {
      sliders,
      width,
      height,
      slidableZoneSize,
      backgroundColor,
      equalColor,
      sliderStyle,
      readOnly
    } = this.props;
    const slidersEqual = this.slidersEqual();
    const sortedSliders = this.sortSliders();

    return (
      <ProgressSlider
        width={width}
        height={height}
        backgroundColor={backgroundColor}
        css={sliderStyle}
        readOnly={readOnly}
        onClick={this.handleSliderClick}
        onMouseDown={this.handleSliderMouseDown}
        onMouseUp={this.handleSliderMouseUp}
        onMouseMove={this.handleSliderMouseMove}
      >
        {sliders.map((slider, i) =>
          <Fragment key={i}>
            <Progress
              className="progress"
              color={slider.color}
              progress={slider.progress}
              height={height}
              equal={slidersEqual}
              equalColor={equalColor}
              zIndex={sortedSliders.indexOf(slider)}
              noTransition={this.state.mouseDown}
            />
            {slider.dot && (
              <Dot
                className="dot"
                right={slider.progress}
                hasIcon={!!slider.dot.icon}
                width={slider.dot.width || 28}
                height={slider.dot.height || 28}
                color={slider.dot.color || slider.color}
                css={slider.dot.style}
                noTransition={this.state.mouseDown}
              >
                {slider.dot.icon && (
                  <DotIcon
                    className="icon"
                    src={slider.dot.icon}
                    width={slider.dot.width || 50}
                    height={slider.dot.height || 50}
                    draggable={false}
                    css={slider.dot.iconStyle}
                  />
                )}
              </Dot>
            )}
          </Fragment>
        )}
        <SlidableZone size={slidableZoneSize} sliderCount={sliders.length} />
      </ProgressSlider>
    );
  };
}
